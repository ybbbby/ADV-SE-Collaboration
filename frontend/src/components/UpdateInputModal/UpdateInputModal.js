import 'date-fns'
import React, { useState } from 'react'
import PropTypes from 'prop-types'
import DateFnsUtils from '@date-io/date-fns'
import {
  Box,
  Button,
  Dialog,
  DialogContent,
  DialogTitle,
  IconButton,
  TextField,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import CloseIcon from '@material-ui/icons/Close'
import {
  MuiPickersUtilsProvider,
  KeyboardDateTimePicker,
} from '@material-ui/pickers'
import LocationSearchInput from '../GoogleMap/LocationSearchInput'

const useStyles = makeStyles((theme) => ({
  closeButton: {
    position: 'absolute',
    right: theme.spacing(1),
    top: theme.spacing(1),
    color: theme.palette.grey[500],
  },
  content: {
    padding: theme.spacing(3),
    textAlign: 'center',
  },
  confirmButton: {
    margin: theme.spacing(1),
    padding: theme.spacing(1.5),
  },
  textInput: {
    width: theme.spacing(50),
  },
  addressInput: {
    width: theme.spacing(50),
    height: theme.spacing(20),
  },
}))

function UpdateInputModal(props) {
  const classes = useStyles()
  const { handleClose, open, type, date, description } = props
  const [selectedDate, setSelectedDate] = useState(date)
  const [newDescription, setNewDescription] = useState(description)
  const [newAddress, setNewAddress] = useState('')
  const [addressError, setAddressError] = useState(false)
  const F = false
  const handleDateChange = (date) => {
    setSelectedDate(date)
  }
  const handleDesChange = (des) => {
    setNewDescription(des.target.value)
  }
  const handleCloseWithNoUpdate = () => {
    handleClose()
    setSelectedDate(date)
    setNewDescription(description)
    setNewAddress('')
  }
  const handleCloseWithUpdate = () => {
    if (type === 1) {
      handleClose(selectedDate)
    } else if (type === 2) {
      handleClose(selectedDate)
    } else if (type === 3) {
      handleClose(newDescription)
    } else if (type === 4) {
      handleClose(newAddress)
      setNewAddress('')
    }
  }
  let selector
  /*
  if (type === 1) {
    selector = (
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <KeyboardDatePicker
          disableToolbar
          variant="inline"
          format="MM/dd/yyyy"
          margin="normal"
          id="date-picker-inline"
          label="Change date"
          value={selectedDate}
          onChange={handleDateChange}
          KeyboardButtonProps={{
            'aria-label': 'change date',
          }}
        />
      </MuiPickersUtilsProvider>
    )
  } else*/
  if (type === 2) {
    selector = (
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <KeyboardDateTimePicker
          id="time-picker"
          label="Change time"
          ampm={F}
          value={selectedDate}
          disablePast
          onChange={handleDateChange}
        />
      </MuiPickersUtilsProvider>
    )
  } else if (type === 3) {
    selector = (
      <Box className={classes.textInput}>
        <TextField
          id="standard-multiline-static"
          label="Change event description"
          multiline
          rows={7}
          fullWidth
          margin="normal"
          value={newDescription}
          onChange={handleDesChange}
          className={classes.descripInput}
          inputProps={{ maxLength: 600 }}
        />
      </Box>
    )
  } else {
    selector = (
      <Box className={classes.addressInput}>
        <LocationSearchInput
          address={newAddress}
          setAddress={setNewAddress}
          error={addressError}
          setError={setAddressError}
        />
      </Box>
    )
  }

  return (
    <Dialog
      open={open}
      onClose={() => handleCloseWithNoUpdate()}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">
        <IconButton
          aria-label="close"
          className={classes.closeButton}
          onClick={() => handleCloseWithNoUpdate()}
        >
          <CloseIcon />
        </IconButton>
      </DialogTitle>
      <DialogContent className={classes.content}>{selector}</DialogContent>
      <Button
        variant="contained"
        color="primary"
        className={classes.confirmButton}
        onClick={handleCloseWithUpdate}
      >
        Confirm
      </Button>
    </Dialog>
  )
}

UpdateInputModal.propTypes = {
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  type: PropTypes.number.isRequired,
  date: PropTypes.object,
  description: PropTypes.string,
}

export default UpdateInputModal
