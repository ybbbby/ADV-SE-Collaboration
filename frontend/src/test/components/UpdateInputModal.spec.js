/* eslint-env jest */
import { shallow } from 'enzyme'
import React from 'react'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import UpdateInputModal from '../../components/UpdateInputModal/UpdateInputModal'
import userEvent from '@testing-library/user-event'

describe('UpdateInputModal', () => {
  test('does not render modal when closed', () => {
    let openModal = false
    const close = jest.fn(() => (openModal = false))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={3}
        description={'Hello'}
      />
    )
    expect(screen.queryByText('Hello')).toBeNull()
  })

  test('renders modal when opened with description', () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={3}
        description={'Hello'}
      />
    )
    expect(screen.queryByText('Hello')).toBeDefined()
  })

  test('renders modal when opened with date', () => {
    let openModal = true
    const date = new Date(Date.UTC(22, 1, 2, 3, 4, 5))
    const close = jest.fn(() => (openModal = false))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={2}
        description={'Hello'}
        date={date}
      />
    )
    expect(screen.queryByText(date.getMonth())).toBeDefined()
  })
  test('renders modal when opened with location', () => {
    let openModal = true
    const date = new Date(Date.UTC(22, 1, 2, 3, 4, 5))
    const close = jest.fn(() => (openModal = false))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={4}
        description={'Hello'}
        date={date}
      />
    )
    expect(screen.queryByText('Search Places ...')).toBeDefined()
  })
  test('close modal', async () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={3}
        description={'Hello'}
      />
    )
    await waitFor(() =>
      userEvent.click(screen.getByRole('button', { name: 'close' }))
    )
    expect(close).toHaveBeenCalledTimes(1)
  })
  test('close modal by foced close', () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={3}
        description={'Hello'}
      />
    )
    fireEvent.keyDown(screen.getByRole('presentation'), {
      key: 'Escape',
      code: 'Escape',
    })
    expect(close).toHaveBeenCalledTimes(1)
  })
  test('change modal with new description', async () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={3}
        description={'Hello'}
      />
    )
    fireEvent.change(screen.getByText('Hello'), {
      target: { value: 'Deku The hero' },
    })
    expect(screen.queryByText('Deku The hero')).toBeDefined()
  })
  test('confirm modal with new date', async () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    const date = new Date(Date.UTC(22, 1, 2, 3, 4, 5))
    render(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={2}
        description={'Hello'}
        date={date}
      />
    )
    await waitFor(() =>
      userEvent.click(screen.getByRole('button', { name: 'Confirm' }))
    )
    expect(close).toHaveBeenCalledTimes(1)
  })
  it('snapshot when description change', () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    const wrapper = shallow(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={3}
        description={'Hello'}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('snapshot when date change', () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    const date = new Date(Date.UTC(22, 1, 2, 3, 4, 5))

    const wrapper = shallow(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={2}
        description={'Hello'}
        date={date}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('snapshot when location change', () => {
    let openModal = true
    const close = jest.fn(() => (openModal = false))
    const date = new Date(Date.UTC(22, 1, 2, 3, 4, 5))

    const wrapper = shallow(
      <UpdateInputModal
        handleClose={() => close()}
        open={openModal}
        type={4}
        description={'Hello'}
        date={date}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
