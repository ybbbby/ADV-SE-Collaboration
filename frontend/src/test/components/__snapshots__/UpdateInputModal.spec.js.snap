// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateInputModal snapshot when date change 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="close"
      className="makeStyles-closeButton-46"
      onClick={[Function]}
    >
      <Memo([object Object]) />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-content-47"
  >
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        allowKeyboardControl={true}
        ampm={false}
        disablePast={true}
        id="time-picker"
        invalidDateMessage="Invalid Date Format"
        label="Change time"
        maxDate={2100-01-01T00:00:00.000Z}
        maxDateMessage="Date should not be after maximal date"
        minDate={1900-01-01T00:00:00.000Z}
        minDateMessage="Date should not be before minimal date"
        onChange={[Function]}
        openTo="date"
        orientation="portrait"
        showTabs={true}
        value={1922-02-02T03:04:05.000Z}
        views={
          Array [
            "year",
            "date",
            "hours",
            "minutes",
          ]
        }
        wider={true}
      />
    </MuiPickersUtilsProvider>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-confirmButton-48"
    color="primary"
    onClick={[Function]}
    variant="contained"
  >
    Confirm
  </WithStyles(ForwardRef(Button))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`UpdateInputModal snapshot when description change 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="close"
      className="makeStyles-closeButton-46"
      onClick={[Function]}
    >
      <Memo([object Object]) />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-content-47"
  >
    <Styled(MuiBox)
      className="makeStyles-textInput-49"
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="standard-multiline-static"
        inputProps={
          Object {
            "maxLength": 600,
          }
        }
        label="Change event description"
        margin="normal"
        multiline={true}
        onChange={[Function]}
        rows={7}
        value="Hello"
      />
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-confirmButton-48"
    color="primary"
    onClick={[Function]}
    variant="contained"
  >
    Confirm
  </WithStyles(ForwardRef(Button))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`UpdateInputModal snapshot when location change 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="close"
      className="makeStyles-closeButton-46"
      onClick={[Function]}
    >
      <Memo([object Object]) />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-content-47"
  >
    <Styled(MuiBox)
      className="makeStyles-addressInput-50"
    >
      <LocationSearchInput
        address=""
        error={false}
        setAddress={[Function]}
        setError={[Function]}
      />
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-confirmButton-48"
    color="primary"
    onClick={[Function]}
    variant="contained"
  >
    Confirm
  </WithStyles(ForwardRef(Button))>
</WithStyles(ForwardRef(Dialog))>
`;
